plugins/radar-android-polar/src/main/java/org/radarbase/passive/polar/PolarManager.kt

```diff
@@ -149,42 +149,92 @@ class PolarManager(
                             streamPpi()
                             streamPpg()
                         }
+
                         else -> {
-                            Log.d(TAG, "No feature was ready")
+                            logger.debug("No feature was ready")
                         }
                     }
                 } else {
-                    Log.d(TAG, "No device was connected")
+                    logger.debug("No device was connected")
                 }
             }
 
             override fun disInformationReceived(identifier: String, uuid: UUID, value: String) {
-                Log.d(TAG, "Firmware: " + identifier + " " + value.trim { it <= ' ' })
+                logger.debug("Firmware: " + identifier + " " + value.trim { it <= ' ' })
             }
 
             override fun batteryLevelReceived(identifier: String, level: Int) {
-                var batteryLevel = level.toFloat() / 100.0f
+                val batteryLevel = level.toFloat() / 100.0f
                 state.batteryLevel = batteryLevel
-                Log.d(TAG, "Battery level $level%, which is $batteryLevel at " + currentTime)
-                send(batteryLevelTopic, PolarBatteryLevel(name, currentTime, currentTime, batteryLevel))
+                logger.debug("Battery level $level%, which is $batteryLevel at $currentTime")
+                send(
+                    batteryLevelTopic,
+                    PolarBatteryLevel(name, currentTime, currentTime, batteryLevel)
+                )
             }
 
         })
 
         try {
-            api.connectToDevice(deviceId)
+            deviceId = service.getPolarDevice()
+            if (deviceId == null) {
+                logger.debug("Searching for Polar devices")
+                connectToPolarDevice()
+            } else {
+                logger.debug("Connecting to Polar device $deviceId")
+                api.connectToDevice(deviceId!!)
+            }
         } catch (a: PolarInvalidArgument) {
             a.printStackTrace()
         }
 
     }
 
-    fun disconnectToPolarSDK(deviceId: String?) {
+    private fun connectToPolarDevice() {
+        autoConnectToPolarSDK()
+    }
+
+    private fun autoConnectToPolarSDK() {
+        if (autoConnectDisposable != null && !autoConnectDisposable!!.isDisposed) {
+            autoConnectDisposable?.dispose()
+        }
+
+        autoConnectDisposable = Flowable.interval(0, 5, TimeUnit.SECONDS)
+            .observeOn(AndroidSchedulers.mainThread())
```

Comment:
I am not very familiar with RxJava, we are using `observeOn(AndroidSchedulers.mainThread())` in every method. It seems that this forces the code inside these methods to always execute on the main thread, which could potentially freeze the UI

Reply:
Yes, that is a good spot; I am also not very familiar with RxJava but this was provided in Polar documentation, hence being used here. I will try to remove it and test it. However, at least here it does not freeze the UI, as it only runs initially until the device is paired. Once there is a paired device, it will always connect to that and will not need scanning of devices. But I agree that sending the data could be an issue.

Reply:
Alternatively, we can use `observerOn` with the main thread and simply wrap the code inside the methods using a handler.

> However, at least in this case, it does not freeze the UI since it only runs initially until the device is paired.

Okay, but there are other functions as well that use `observeOn(AndroidSchedulers.mainThread())`.